contracts/Zauction.sol

```diff
@@ -93,33 +97,103 @@ contract ZAuction is Initializable, OwnableUpgradeable {
-    // Bidder -> Owner, send funds
-    SafeERC20.safeTransferFrom(token, bidder, msg.sender, bid - royalty);
-    // Bidder -> Registrar, pay royalty
-    SafeERC20.safeTransferFrom(
-      token,
-      bidder,
-      registrar.minterOf(tokenId),
-      royalty
-    );
+
+    // Transfer payment and royalties
+    paymentTransfers(bidder, bid, msg.sender, rootId, tokenId);
+
     // Owner -> Bidder, send NFT
-    nftContract.safeTransferFrom(msg.sender, bidder, tokenId);
+    nft.safeTransferFrom(msg.sender, bidder, tokenId);
+
     emit BidAccepted(
       auctionId,
       bidder,
       msg.sender,
       bid,
-      address(nftContract),
+      address(nft),
       tokenId,
       expireBlock
     );
   }
 
+  function paymentTransfers(
+    address bidder,
+    uint256 bid,
+    address sender,
+    uint256 rootId,
+    uint256 tokenId
+  ) public {
+    address rootOwner = nft.ownerOf(rootId);
+    uint256 rootRoyalty = calculateRootOwnerRoyalty(rootId, bid, tokenId);
+    uint256 minterRoyalty = calculateMinterRoyalty(bid, tokenId);
+
+    uint256 bidActual = bid - minterRoyalty - rootRoyalty;
+
+    // Bidder -> Owner, pay transaction
+    SafeERC20.safeTransferFrom(token, bidder, sender, bidActual);
+
+    // Bidder -> Minter, pay minter royalty
+    SafeERC20.safeTransferFrom(
+      token,
+      bidder,
+      registrar.minterOf(tokenId),
+      minterRoyalty
+    );
+
+    // Bidder -> Root Owner, pay root owner royalty
+    SafeERC20.safeTransferFrom(token, bidder, rootOwner, rootRoyalty);
+  }
+
+  function calculateRootOwnerRoyalty(
+    uint256 rootId,
+    uint256 bid,
+    uint256 id
+  ) public view returns (uint256) {
+    require(id > 0, "zAuction: must provide a valid id");
+
+    // Find what percent they've specified as a royalty
+    uint256 royalty = rootDomainRoyalty[rootId];
+
+    // If not found or 0% royalty
+    if (royalty == 0) return 0;
+    // if not found because it's new do we just say 0?
+
+    // Pad with 18 zeroes for precision
+    uint256 divisor = (100 * 10**18) / (royalty * 10**18);
+    uint256 calculatedRoyalty = (bid / divisor);
+    return calculatedRoyalty;
+  }
+
+  function calculateMinterRoyalty(uint256 bid, uint256 id)
+    public
+    view
+    returns (uint256)
+  {
+    require(id > 0, "zAuction: must provide a valid id");
+    // Returns with 5 decimal points of accuracy
+    uint256 domainRoyalty = registrar.domainRoyaltyAmount(id);
+    if (domainRoyalty == 0) return 0; // same here, 0 for 0% or 0 for not found?
+
+    // Pad with 18 zeroes for precision, 13 for domainRoyalty as already 5 decimal points captured
+    uint256 divisor = (100 * (10**18)) / (domainRoyalty * 10**13);
+    uint256 calculatedRoyalty = (bid / divisor);
+    return calculatedRoyalty;
+  }
+
+  function setRootRoyaltyAmount(uint256 id, uint256 amount) public {
+    rootDomainRoyalty[id] = amount;
+  }
+
+  function setRegistrarRoyaltyAmount(uint256 id, uint256 amount) public {
```

Comment:
While this would be nice to have, this won't actually work since `registrar.setDomainRoyaltyAmount` requires that the caller own the domain, and in this case the caller would be the `zAuction` contract which doesn't own the domain. Therefore, this function should be removed.